File: calkin_wilf/calkin_wilf.cpp
#include <iostream>
#include <vector>
#include <boost/dynamic_bitset.hpp>
#include <boost/rational.hpp>

using rational = boost::rational<unsigned long>;
using bitset = boost::dynamic_bitset<unsigned long>;

unsigned long floor(const rational& r) {
    return r.numerator()/r.denominator();
}

class calkin_wilf {
public:
    rational next() {
        rational term = term_;
        term_ = 1UL/(2UL * floor(term) + 1UL - term);
        return term;
    }
private:
    rational term_ = 0;
};

std::vector<unsigned long> continued_fraction(const rational& r) {
    unsigned long a = r.numerator();
    unsigned long b = r.denominator();
    std::vector<unsigned long> result;
    for (;;) {
        result.push_back(a/b);
        unsigned long c = a;
        a = b;
        b = c % b;
        if (a == 1)
            break;
    }
    if (result.size() > 0 && result.size() % 2 == 0) {
        --result.back();
        result.push_back(1);
    }
    return result;
}

unsigned long term_number(const rational& r) {
    bitset b;
    bool d = true;
    for (unsigned long n : continued_fraction(r)) {
        for (unsigned long i = 0; i < n; ++i)
            b.push_back(d);
        d = !d;
    }
    return b.to_ulong();
}

int main() {
    calkin_wilf cf;
    std::cout << "First 21 terms of the Calkin-Wilf sequence are:\n";
    for (int i = 0; i <= 20; ++i) {
        rational term = cf.next();
        std::cout << std::setw(2) << i << ": " << term << '\n';
    }
    rational r(83116, 51639);
    std::cout << r << " is the " << term_number(r) << "th term of the sequence.\n";
}

File: calkin_wilf/calkin_wilf.rs
// [dependencies]
// num-rational = "0.3"

use num::rational::Rational;

fn calkin_wilf_next(term: &Rational) -> Rational {
    Rational::from_integer(1)/(Rational::from_integer(2) * term.floor() + 1 - term)
}

fn continued_fraction(r: &Rational) -> Vec<isize> {
    let mut a = *r.numer();
    let mut b = *r.denom();
    let mut result = Vec::new();
    loop {
        result.push(a/b);
        let c = a;
        a = b;
        b = c % b;
        if a == 1 {
            break;
        }
    }
    let len = result.len();
    if len != 0 && len % 2 == 0 {
        result[len - 1] -= 1;
        result.push(1);
    }
    result
}

fn term_number(r: &Rational) -> usize {
    let mut result: usize = 0;
    let mut d: usize = 1;
    let mut p: usize = 0;
    for n in continued_fraction(r) {
        for _ in 0..n {
            result |= d << p;
            p += 1;
        }
        d ^= 1;
    }
    result
}

fn main() {
    println!("First 21 terms of the Calkin-Wilf sequence are:");
    let mut term = Rational::from_integer(0);
    for i in 0..=20 {
        println!("{:2}: {}", i, term);
        term = calkin_wilf_next(&term);
    }
    let r = Rational::new(83116, 51639);
    println!("{} is the {}th term of the sequence.", r, term_number(&r));
}
